applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+
+void AdjointFiniteDifferencingBaseElement::CalculateSensitivityMatrix(const Variable<double>& rDesignVariable, Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    // define working variables
+    Vector RHS_undist;
+    Vector RHS_dist;
+    ProcessInfo copy_process_info = rCurrentProcessInfo;
+
+    // Compute RHS before disturbing
+    mpPrimalElement->CalculateRightHandSide(RHS_undist, copy_process_info);
+    rOutput.resize(1,RHS_undist.size());
+
+    // Get disturbance measure
+    double delta = this->GetValue(DISTURBANCE_MEASURE);
+    double correction_factor = this->GetDisturbanceMeasureCorrectionFactor(rDesignVariable);
+    delta *= correction_factor;
+
+    if ( mpPrimalElement->GetProperties().Has(rDesignVariable) )
+    {
+        // Save properties and its pointer
+        Properties& r_global_property = mpPrimalElement->GetProperties();
+        Properties::Pointer p_global_properties = mpPrimalElement->pGetProperties();
+
+        // Create new property and assign it to the element
+        Properties::Pointer p_local_property(new Properties(r_global_property));
+        mpPrimalElement->SetProperties(p_local_property);
+
+        // Disturb the design variable
+        const double current_property_value = mpPrimalElement->GetProperties()[rDesignVariable];
+        p_local_property->SetValue(rDesignVariable, (current_property_value + delta));
+
+        this->AfterPerturbation(rDesignVariable, rCurrentProcessInfo);
+
+        // Compute RHS after disturbance
+        mpPrimalElement->CalculateRightHandSide(RHS_dist, copy_process_info);
+
+        // Compute derivative of RHS w.r.t. design variable with finite differences
+        noalias(RHS_dist) -= RHS_undist;
+        RHS_dist /= delta;
+        for(unsigned int i = 0; i < RHS_dist.size(); i++)
+            rOutput(0, i) = RHS_dist[i];
+
+        // Give element original properties back
+        mpPrimalElement->SetProperties(p_global_properties);
+
+        this->AfterPerturbation(rDesignVariable, rCurrentProcessInfo);
+
+        //call one last time to make sure everything is as it was before TODO improve this..
+        mpPrimalElement->CalculateRightHandSide(RHS_dist, copy_process_info);
+    }
+    else
+        rOutput.clear();
+
+    KRATOS_CATCH("")
+
+}
+
+void AdjointFiniteDifferencingBaseElement::CalculateSensitivityMatrix(const Variable<array_1d<double,3>>& rDesignVariable, Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    // define working variables
+    Vector RHS_undist;
+    Vector RHS_dist;
+    ProcessInfo copy_process_info = rCurrentProcessInfo;
+
+    // Get disturbance measure
+    double delta= this->GetValue(DISTURBANCE_MEASURE);
+    double correction_factor = this->GetDisturbanceMeasureCorrectionFactor(rDesignVariable);
+    delta *= correction_factor;
+
+    if(rDesignVariable == SHAPE)
```

Comment:
I would be one  of the dev Mike mentioned, I was having a hard time understanding this

Reply:
The reasons behind this are the following:
1. The variable is shape (or coordinate) and not shape_sensitivity
2. For all other design variables (the ones that have a corresponding variable, like e.g. THICKNESS or other properties), the variable can be directly used to do the perturbation. The corresponding sensitivities are then written on the variable_name+"_SENSITIVITY". Shape or node coordinates are slightly different here, because they are not defined via a variable.

Reply:
Still open